### PR TITLE
feat(ZN-390): add abstraction over use navigate and use params

### PR DESCRIPTION
**Pull request by kacper-witas-tsh**
useParams takes generic which is route, for example `useParams<'foo/:bar'>()`, it throws an error if bar is not defined because it means that given component was rendered at route without expected `bar` in params. It also infers the type of params from the string, for example:

```tsx
const { baz, bar } = useParams<'foo/:baz/:bar'>(); // baz: string, bar: string
```

useNavigate returns a function which infers the type of params from url string, for example:

```tsx
const navigate = useNavigate();

navigate('foo/:bar', { params: { bar: 'baz' } }); // ts error if you omit the parameters
navigate('foo/:bar', { params: { bar: 'baz' }, replace: true }); // it also preserves types of options object expected by original navigate function
```

This PR also includes adding extra feature into custom test renderer, now it accepts router configuration as shown in `src/app/user/User.test.tsx`

### CODE REVIEW

File: src/hooks/useNavigate/useNavigate.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  ) => {
+    let parsedPath: string = path;
+    if (options && 'params' in options) {
+      Object.entries(options.params).forEach(([key, value]) => {
+        parsedPath = parsedPath.replace(`:${key}`, value as string);
+      });
+    }
+
+    navigate(parsedPath);
+  };
```
Comment (barto-tsh) [inline src/hooks/useNavigate/useNavigate.ts:21]: what do you think about making this part immutable?
```suggestion
    if (options && 'params' in options) {
      return navigate(Object.entries(options.params).reduce(([key, value]) => {
  return parsedPath.replace(`:${key}`, value);
}, path))
    }

    navigate(path);
  };
```

File: src/hooks/useParams/useParams.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { useParams as ReactUseParams } from 'react-router-dom';
+
+import { PathParams } from 'types/params-utils';
+import { Simplify } from 'types/simplify';
+
+function assertParamsAreValid<Path extends string>(
```
Comment (barto-tsh) [inline src/hooks/useParams/useParams.ts:6]: could you convert it to lambda?

File: src/tests/index.tsx [hunk 3]
```diff
@@ -10,8 +10,48 @@ import { defaultLocale } from 'context/locale/defaultLocale';
...
+  | { withRouter: false };
+
+type ExtraRenderOptions = {
+  routerConfig?: RouterConfig;
+};
+
+type WrapperProps = {
+  children: ReactNode;
+  routerConfig?: RouterConfig;
+};
```
Comment (barto-tsh) [inline src/tests/index.tsx:30]: you could move those types to separate file 😁

File: src/tests/index.tsx [hunk 4]
```diff
@@ -30,20 +70,29 @@ const Wrapper = ({ children }: { children?: ReactNode }) => {
...
+  options: RenderOptions<Q> & ExtraRenderOptions,
+): RenderResult<Q>;
 function customRender<Q extends Queries>(
   ui: React.ReactElement,
-  options?: RenderOptions<Q> | Omit<RenderOptions, 'queries'>,
+  options?: (RenderOptions<Q> | Omit<RenderOptions, 'queries'>) & ExtraRenderOptions,
 ): RenderResult<Q> | RenderResult {
+  function Wrapper({ children }: Pick<WrapperProps, 'children'>) {
+    return <_Wrapper routerConfig={options?.routerConfig}>{children}</_Wrapper>;
+  }
```
Comment (barto-tsh) [inline src/tests/index.tsx:95]: those could be converted to lambda notation 😁

**Review by barto-tsh — APPROVED** [review]

**Comment by barto-tsh** [comment]
looks really good! 😁

File: src/hooks/useParams/useParams.ts (hunk 2 above)
Comment (kacper-witas-tsh) [reply src/hooks/useParams/useParams.ts:6]: I think functions which are assertions have to be named function definitions 🤔 not sure if I am right

File: src/hooks/useNavigate/useNavigate.ts (hunk 1 above)
Comment (kacper-witas-tsh) [reply src/hooks/useNavigate/useNavigate.ts:21]: nice one, I think your reduce definition lacks accumulator, I've added it 🙂

File: src/tests/index.tsx (hunk 3 above)
Comment (kacper-witas-tsh) [reply src/tests/index.tsx:30]: thanks, I forgot

File: src/tests/index.tsx (hunk 4 above)
Comment (kacper-witas-tsh) [reply src/tests/index.tsx:95]: I thought about error's stack trace, I think lambda component won't have display name 🤔 I might be wrong tho.
Anyway getting error in this part of code shouldn't happen too often so I'll make it as a lambda as you suggested

File: src/hooks/useNavigate/useNavigate.ts (hunk 1 above)
Comment (barto-tsh) [reply src/hooks/useNavigate/useNavigate.ts]: ahh right, my bad 😅

File: src/hooks/useParams/useParams.ts (hunk 2 above)
Comment (barto-tsh) [reply src/hooks/useParams/useParams.ts:6]: you're right ☝️

**Review by barto-tsh — APPROVED** [review]

**Review by szymonchudy — APPROVED** [review]

**Review by dawidkostrzewa — APPROVED** [review]

**Pull request merged**